chainntnfs/txconfnotifier.go

```diff
@@ -0,0 +1,227 @@
+// (or be notified immediately if the tx already does), and the TxConfNotifier
+// will watch changes to the blockchain in order to satisfy these requests.
+type TxConfNotifier struct {
+	// currentHeight is the height of the tracked blockchain. It is used to
+	// determine the number of confirmations a tx has and ensure blocks are
+	// connected and disconnected in order.
+	currentHeight uint32
+
+	// reorgSafetyLimit is the chain depth beyond which it is assumed a block
+	// will not be reorganized out of the chain. This is used to determine when
+	// to prune old confirmation requests so that reorgs are handled correctly.
+	// The coinbase maturity period is a reasonable value to use.
+	reorgSafetyLimit uint32
+
+	// confNotifications is an index of notification requests by transaction
+	// hash.
+	confNotifications map[chainhash.Hash][]*ConfNtfn
+
+	// confTxsByInitialHeight is an index of watched transactions by the height
+	// that they are included at in the blockchain. This is tracked so that
+	// incorrect notifications are not sent if a transaction is reorganized out
+	// of the chain and so that negative confirmations can be recognized.
+	confTxsByInitialHeight map[uint32][]*chainhash.Hash
+
+	// ntfnsByConfirmHeight is an index of notification requests by the height
+	// at which the transaction will have sufficient confirmations.
+	ntfnsByConfirmHeight map[uint32][]*ConfNtfn
+}
+
+// NewTxConfNotifier creates a TxConfNotifier. The current height of the
+// blockchain is accepted as a parameter.
+func NewTxConfNotifier(startHeight uint32, reorgSafetyLimit uint32) *TxConfNotifier {
+	return &TxConfNotifier{
+		currentHeight:          startHeight,
+		reorgSafetyLimit:       reorgSafetyLimit,
+		confNotifications:      make(map[chainhash.Hash][]*ConfNtfn),
+		confTxsByInitialHeight: make(map[uint32][]*chainhash.Hash),
+		ntfnsByConfirmHeight:   make(map[uint32][]*ConfNtfn),
+	}
+}
+
+// Register handles a new notification request. The client will be notified when
+// the transaction gets a sufficient number of confirmations on the blockchain.
+// If the transaction has already been included in a block on the chain, the
+// confirmation details must be given as the txConf argument, otherwise it
+// should be nil. If the transaction already has the sufficient number of
+// confirmations, this dispatches the notification immediately.
+func (tcn *TxConfNotifier) Register(ntfn *ConfNtfn, txConf *TxConfirmation) {
+	if txConf == nil || txConf.BlockHeight > tcn.currentHeight {
+		// Transaction is unconfirmed.
+		tcn.confNotifications[*ntfn.TxID] =
+			append(tcn.confNotifications[*ntfn.TxID], ntfn)
+		return
+	}
+
+	// If the transaction already has the required confirmations, dispatch
+	// notification immediately, otherwise record along with the height at
+	// which to notify.
+	confHeight := txConf.BlockHeight + ntfn.NumConfirmations - 1
+	if confHeight <= tcn.currentHeight {
+		ntfn.Event.Confirmed <- txConf
```

Comment:
This should likely select on a quit channel to avoid the possibility of a deadlock.

Reply:
Perhaps, though the documentation of `ConfirmationEvent` specifies that this channel [MUST be buffered](https://github.com/lightningnetwork/lnd/blob/master/chainntnfs/interface.go#L101), in which case this operation cannot block.

Reply:
Buffered to what extent though? What if a big re-org happens and 30 blocks are instantly connected? 

The implementation should be defensive and ensure that there's _no way_ it can cause a deadlock.

Reply:
See this logic: https://github.com/lightningnetwork/lnd/pull/410/files#diff-292e14dd3a6f7f638d365bc802dac419R207. I believe that makes it safe to have a channel capacity of 1.

Reply:
Independent of the scenario above (alternatively the caller stops reading, 30 blocks being added on). There's no reason to _not_ add a quit channel. It's defensive programming. We've had several cases of deadlocks occurring due to unconditional sends on channels. As a result, we've started to adopt a best practice policy of _always_ do conditional sends in order to rule out any possible deadlocks.

Reply:
Makes sense. Added.

Reply:
Master was recently updated to _always_ log whenever a confirmation is dispatched. Ideally this patch replicates the existing set of logs _exactly_ to mirror the prior operation.